File: backend/app/schemas/lms.py
"""LMS student-facing schemas (LM1-3).

The content models are the *second* enforcement of the §2 answer-leakage
guarantee. Every model is `extra="forbid"`, so a payload that retains
`is_correct` or `explanation` fails response validation instead of silently
passing — the leak the service-layer `student_view` already strips is caught
again at the response boundary (plan: "Pydantic response models on the student
routes enforce it a second time").
"""

from datetime import date, datetime
from typing import Literal, Union
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.schemas.curriculum import PrerequisiteItemOut


class ContentText(BaseModel):
    model_config = ConfigDict(extra="forbid")
    body: str


class QuizOptionOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    text: str


class QuizQuestionOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    prompt: str
    options: list[QuizOptionOut]


class ContentQuiz(BaseModel):
    model_config = ConfigDict(extra="forbid")
    pass_threshold: int
    questions: list[QuizQuestionOut]


class FlashcardCardOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    term: str
    definition: str


class ContentFlashcards(BaseModel):
    model_config = ConfigDict(extra="forbid")
    title: str | None = None
    cards: list[FlashcardCardOut]


class ContentVideo(BaseModel):
    model_config = ConfigDict(extra="forbid")
    transcode_status: str | None = None
    duration_seconds: int | None = None


class ContentAttachment(BaseModel):
    model_config = ConfigDict(extra="forbid")
    filename: str | None = None
    size_bytes: int | None = None


class ContentMission(BaseModel):
    """P5-5 — a mission embedded in a course module. `mission_title`/
    `mission_kind`/`points`/`attempt_status` are enriched by the router
    (mirrors ContentVideo's transcode_status — not stored in `content`
    itself, looked up fresh per request)."""
    model_config = ConfigDict(extra="forbid")
    mission_id: UUID
    variant_id: UUID | None = None
    mission_title: str | None = None
    mission_kind: str | None = None
    points: int | None = None
    attempt_status: str | None = None


ModuleContent = Union[ContentQuiz, ContentText, ContentFlashcards, ContentVideo, ContentAttachment, ContentMission]


# ── video checkpoints (timeline notes + mid-video quizzes, 2026-08-07) ───────

class CheckpointNoteOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    body: str


class CheckpointQuizOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    question_type: Literal["mcq", "multiselect", "open"]
    prompt: str
    # None for open questions — nothing to choose from.
    options: list[QuizOptionOut] | None = None


CheckpointContent = Union[CheckpointNoteOut, CheckpointQuizOut]


class VideoCheckpointOut(BaseModel):
    id: UUID
    start_seconds: int
    end_seconds: int | None = None
    kind: Literal["note", "quiz"]
    content: CheckpointContent


class CheckpointAnswerIn(BaseModel):
    # int (mcq), list[int] (multiselect), or str (open) — validated against
    # the checkpoint's actual question_type server-side, same "don't trust
    # the client's own shape" posture as QuizAnswersIn.
    answer: int | list[int] | str


class CheckpointAnswerOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    # None for open questions (not graded) or if the student skipped.
    correct: bool | None = None
    explanation: str | None = None


# ── catalog / course outline (login-only) ───────────────────────────────────

class CourseCatalogOut(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    kind: str
    image_url: str | None = None
    level: str | None = None
    track: str | None = None
    access_mode: str = "open"  # P1-7 — open|invite|paid
    enrolled: bool = False  # this caller's own active, unexpired enrollment
    locked: bool = False  # 7B-2 — unmet prerequisites; independent of access_mode


# ── my-courses dashboard (student, LMS redesign 2026-08-06) ────────────────

class DashboardStatsOut(BaseModel):
    in_progress: int
    total_enrolled: int
    modules_done: int


class ResumePointerOut(BaseModel):
    course_id: UUID
    course_title: str
    module_id: UUID
    module_title: str
    next_item_id: UUID | None = None
    mandatory_completed: int
    mandatory_total: int


class DashboardCourseOut(BaseModel):
    course_id: UUID
    title: str
    kind: str
    status: str  # not_started | in_progress | completed
    modules_done: int
    modules_total: int
    pct: int


class MyCoursesOut(BaseModel):
    stats: DashboardStatsOut
    resume: ResumePointerOut | None = None
    courses: list[DashboardCourseOut]


class ActivityItemOut(BaseModel):
    item_id: UUID
    item_title: str | None = None
    item_kind: str
    course_id: UUID
    course_title: str
    completed_at: datetime | None = None


class ModuleLockOut(BaseModel):
    module_id: UUID
    title: str | None = None
    position: int
    locked: bool
    mandatory_total: int
    mandatory_completed: int


class CourseDetailOut(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    kind: str
    enrolled: bool
    completed: bool
    modules: list[ModuleLockOut]
    image_url: str | None = None
    outcomes: list[str] = []
    level: str | None = None
    track: str | None = None
    instructor_name: str | None = None
    instructor_title: str | None = None
    instructor_photo_url: str | None = None
    access_mode: str = "open"  # P1-7 — open|invite|paid, drives the CTA
    locked: bool = False  # 7B-2 — unmet prerequisites; independent of access_mode
    prerequisites: list[PrerequisiteItemOut] = []
    # Stage S — integer minor units, matching Stripe's unit_amount. NULL
    # unless access_mode == "paid".
    price_cents: int | None = None
    currency: str = "usd"


# ── enrollment ──────────────────────────────────────────────────────────────

class EnrollIn(BaseModel):
    course_id: UUID


class EnrollmentOut(BaseModel):
    id: UUID
    course_id: UUID
    source: str
    status: str
    created_at: datetime | None = None
    expires_at: datetime | None = None  # P1-3 — NULL means perpetual


# ── checkout (Stage S, Stripe) ───────────────────────────────────────────────

class CheckoutSessionOut(BaseModel):
    checkout_url: str


class CheckoutFulfillOut(BaseModel):
    status: str  # pending | paid | refunded | disputed | failed
    course_id: UUID


# ── module read (enrolled student only) ─────────────────────────────────────

class ModuleItemOut(BaseModel):
    id: UUID
    kind: Literal["video", "text", "quiz", "flashcards", "attachment", "mission"]
    title: str | None = None
    position: int
    content: ModuleContent
    # this student's own per-item state, so the player can render progress
    status: str | None = None


class ModuleOut(BaseModel):
    id: UUID
    course_id: UUID
    title: str
    position: int
    items: list[ModuleItemOut]


# ── progress + quiz submission ──────────────────────────────────────────────

class ProgressIn(BaseModel):
    action: Literal["video-watched", "text-viewed", "quiz-attempt", "flashcards-skipped", "attachment-viewed"]


class ProgressOut(BaseModel):
    status: str
    quiz_attempts: int
    best_score: float | None = None
    completed_at: datetime | None = None


class QuizAnswersIn(BaseModel):
    answers: list[int]


class QuizAnswerCheckIn(BaseModel):
    question_index: int
    answer: int


class QuizAnswerCheckOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    correct: bool
    explanation: str | None = None
    correct_text: str | None = None


class AttachmentUrlOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    url: str
    filename: str | None = None


class QuizReviewItemOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    prompt: str | None = None
    selected: int
    correct: bool
    explanation: str | None = None
    correct_text: str | None = None


class QuizReviewOut(BaseModel):
    model_config = ConfigDict(extra="forbid")
    score: float
    passed: bool
    pass_threshold: int
    attempts: int
    best_score: float | None = None
    questions: list[QuizReviewItemOut]


# ── learning paths (self-paced ordered course sequences, 2026-08-08) ───────

class LearningPathCatalogOut(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    image_url: str | None = None
    course_count: int
    mission_count: int
    total_duration_seconds: int
    pct: int
    # 2026-08-12 — "am I on this path", so the landing page can split
    # "your paths" from "explore". True iff the caller has an active
    # enrollment in at least one of the path's courses; `pct` alone can't
    # answer it (a freshly-started path is 0% but is still yours), which is
    # the same distinction `CourseCatalogOut.enrolled` already draws.
    enrolled: bool = False


class MyCertificateOut(BaseModel):
    """One earned LMS certificate (2026-08-13). `url` is signed at query
    time from bucket/file_path, the resolve_url pattern — never stored."""
    id: UUID
    type: str  # lms_course_completion | lms_path_completion
    title: str  # the course or learning path it was earned for
    course_id: UUID | None = None
    learning_path_id: UUID | None = None
    issued_at: datetime | None = None
    url: str | None = None


class LearningPathStepOut(BaseModel):
    position: int
    course_id: UUID
    title: str
    kind: str
    state: Literal["done", "current", "mission", "locked"]
    pct: int
    modules_done: int
    modules_total: int


class LearningPathDetailOut(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    image_url: str | None = None
    pct: int
    course_count: int
    mission_count: int
    total_duration_seconds: int
    steps: list[LearningPathStepOut]


# ── leaderboard (P2-4) — not wired into any student-facing page yet, D6 is
# still an open operator decision (PHASE2_EXECUTION_PLAN.md §2) ────────────

class LeaderboardEntryOut(BaseModel):
    rank: int
    user_id: UUID
    display_name: str
    points: int


# ── my-programs (P4-3) — the cohort view a student cannot currently see ────

class MyProgramCourseOut(BaseModel):
    course_id: UUID
    title: str
    enrolled: bool
    progress_pct: int


class MyProgramOut(BaseModel):
    registration_id: UUID
    cohort_id: UUID
    program_name: str
    cohort_name: str
    starts_on: date | None = None
    ends_on: date | None = None
    location_name: str | None = None
    location_address: str | None = None
    instructor_name: str | None = None
    attended_sessions: int
    total_sessions: int
    courses: list[MyProgramCourseOut] = []
    missions: list = []  # Stage 5 fills this in; empty, not omitted, until then


# ── LMS Program checklist (2026-08-21 redesign) ─────────────────────────────

LmsProgramItemStatus = Literal["pending", "done", "awaiting_confirmation"]


class LmsProgramChecklistItemOut(BaseModel):
    id: UUID
    position: int
    item_type: str
    title: str
    description: str | None = None
    optional: bool
    requires_confirmation: bool
    status: LmsProgramItemStatus
    # Resolved link, whatever the item type: a course id, a mission
    # attempt id (the ops-assigned run itself, never the mission catalog
    # entry), or the stored external_url — the frontend picks which by
    # `item_type`, same shape the Poster tab already established.
    course_id: UUID | None = None
    mission_attempt_id: UUID | None = None
    # mission_id/mission_kind: which route a mission_run item's "Continue"
    # link needs — design/operate attempts have their own attempt-keyed
    # page, every other kind is viewed inline on the mission's own page.
    mission_id: UUID | None = None
    mission_kind: str | None = None
    external_url: str | None = None
    submission_prompt: str | None = None
    # What the student themself submitted back, for a `submission` item.
    submitted_url: str | None = None


class LmsProgramAssignmentSummaryOut(BaseModel):
    assignment_id: UUID
    lms_program_id: UUID
    name: str
    cohort_id: UUID | None = None
    cohort_name: str | None = None
    items_total: int
    items_done: int
    pct: int
    next_item_title: str | None = None
    certificate_required: bool
    certificate_earned: bool


class LmsProgramChecklistOut(BaseModel):
    assignment_id: UUID
    lms_program_id: UUID
    name: str
    description: str | None = None
    cohort_id: UUID | None = None
    cohort_name: str | None = None
    certificate_required: bool
    certificate_earned: bool
    items: list[LmsProgramChecklistItemOut]


class LmsProgramItemSubmitIn(BaseModel):
    url: str

File: frontend/src/api/lms.ts
/** LMS student API (LM1-8) — thin wrappers over `/lms/*` (LM1-3) and the
 * video token/stream routes (LM1-6). Types mirror `schemas/lms.py` field for
 * field; `content` shapes match LMS_EXECUTION_PLAN.md §2's four kinds.
 */
import { api } from "./client";

export interface CourseCatalogItem {
  id: string;
  title: string;
  description: string | null;
  kind: "course" | "mission";
  image_url: string | null;
  level: "beginner" | "intermediate" | "advanced" | null;
  track: string | null;
  access_mode: "open" | "invite" | "paid"; // P1-7
  enrolled: boolean;
  locked: boolean; // 7B-2 — unmet prerequisites; independent of access_mode
}

/** 7B-2 — a unified course/mission prerequisite edge, evaluated for the
 * caller. `item_type: "mission"` has no course-side page to link to, so
 * it's rendered as plain text here (mirrors how a course prerequisite on a
 * mission's own page has no mission-side page to link to either). */
export interface CoursePrerequisite {
  item_type: "course" | "mission";
  item_id: string;
  title: string;
  satisfied: boolean;
}

export interface ModuleLock {
  module_id: string;
  title: string | null;
  position: number;
  locked: boolean;
  mandatory_total: number;
  mandatory_completed: number;
}

export interface CourseDetail {
  id: string;
  title: string;
  description: string | null;
  kind: "course" | "mission";
  enrolled: boolean;
  completed: boolean;
  modules: ModuleLock[];
  image_url: string | null;
  outcomes: string[];
  level: "beginner" | "intermediate" | "advanced" | null;
  track: string | null;
  instructor_name: string | null;
  instructor_title: string | null;
  instructor_photo_url: string | null;
  access_mode: "open" | "invite" | "paid"; // P1-7
  locked: boolean; // 7B-2 — unmet prerequisites; independent of access_mode
  prerequisites: CoursePrerequisite[];
  // Stage S (Stripe Checkout) — integer minor units, matching Stripe's
  // unit_amount. Null unless access_mode === "paid".
  price_cents: number | null;
  currency: string;
}

export interface QuizOption {
  text: string;
}

export interface QuizQuestion {
  prompt: string;
  options: QuizOption[];
}

export type ModuleItemContent =
  | { body: string } // text
  | { pass_threshold: number; questions: QuizQuestion[] } // quiz
  | { title: string | null; cards: { term: string; definition: string }[] } // flashcards
  | { transcode_status: string | null; duration_seconds: number | null } // video
  | { filename: string | null; size_bytes: number | null } // attachment
  | {
      // mission (P5-5) — mission_title/mission_kind/points/attempt_status
      // are enriched by the backend per request, not authored content.
      mission_id: string;
      variant_id: string | null;
      mission_title: string | null;
      mission_kind: string | null;
      points: number | null;
      attempt_status: "in_progress" | "submitted" | "passed" | "failed" | "abandoned" | null;
    };

export interface ModuleItem {
  id: string;
  kind: "video" | "text" | "quiz" | "flashcards" | "attachment" | "mission";
  title: string | null;
  position: number;
  content: ModuleItemContent;
  status: "not_started" | "in_progress" | "completed" | "skipped" | null;
}

export interface ModuleDetail {
  id: string;
  course_id: string;
  title: string;
  position: number;
  items: ModuleItem[];
}

export interface QuizReviewQuestion {
  prompt: string | null;
  selected: number;
  correct: boolean;
  explanation: string | null;
  correct_text: string | null;
}

export interface QuizReview {
  score: number;
  passed: boolean;
  pass_threshold: number;
  attempts: number;
  best_score: number | null;
  questions: QuizReviewQuestion[];
}

export interface QuizAnswerCheck {
  correct: boolean;
  explanation: string | null;
  correct_text: string | null;
}

export interface ProgressResult {
  status: string;
  quiz_attempts: number;
  best_score: number | null;
  completed_at: string | null;
}

// ── video checkpoints (timeline notes + mid-video quizzes, 2026-08-07) ─────

export type CheckpointQuestionType = "mcq" | "multiselect" | "open";

export type CheckpointContent =
  | { body: string } // note
  | { question_type: CheckpointQuestionType; prompt: string; options: QuizOption[] | null }; // quiz

export interface VideoCheckpoint {
  id: string;
  start_seconds: number;
  end_seconds: number | null;
  kind: "note" | "quiz";
  content: CheckpointContent;
}

export interface CheckpointAnswerResult {
  correct: boolean | null;
  explanation: string | null;
}

// ── dashboard (LMS redesign, 2026-08-06) ────────────────────────────────────

export interface DashboardStats {
  in_progress: number;
  total_enrolled: number;
  modules_done: number;
}

export interface ResumePointer {
  course_id: string;
  course_title: string;
  module_id: string;
  module_title: string;
  next_item_id: string | null;
  mandatory_completed: number;
  mandatory_total: number;
}

export interface DashboardCourse {
  course_id: string;
  title: string;
  kind: "course" | "mission";
  status: "not_started" | "in_progress" | "completed";
  modules_done: number;
  modules_total: number;
  pct: number;
}

export interface MyCourses {
  stats: DashboardStats;
  resume: ResumePointer | null;
  courses: DashboardCourse[];
}

export async function fetchMyCourses(): Promise<MyCourses> {
  const { data } = await api.get<MyCourses>("/lms/my-courses");
  return data;
}

export interface ActivityItem {
  item_id: string;
  item_title: string | null;
  item_kind: "video" | "text" | "quiz" | "flashcards";
  course_id: string;
  course_title: string;
  completed_at: string | null;
}

export async function fetchMyActivity(): Promise<ActivityItem[]> {
  const { data } = await api.get<ActivityItem[]>("/lms/my-activity");
  return data;
}

// ── learning paths (self-paced ordered course sequences, 2026-08-08) ───────

export interface LearningPathCatalogItem {
  id: string;
  title: string;
  description: string | null;
  image_url: string | null;
  course_count: number;
  mission_count: number;
  total_duration_seconds: number;
  pct: number;
  /** True iff you have an active enrollment in at least one of the path's
   * courses — `pct` can't answer this (a just-started path is still 0%). */
  enrolled: boolean;
  // Bundle pricing (2026-08-21) — null price_cents means not purchasable.
  price_cents: number | null;
  currency: string;
  /** True iff you already have an active enrollment in every step's course —
   * a bundle purchase would have nothing left to grant. */
  fully_owned: boolean;
}

export interface LearningPathStep {
  position: number;
  course_id: string;
  title: string;
  kind: "course" | "mission";
  state: "done" | "current" | "mission" | "locked";
  pct: number;
  modules_done: number;
  modules_total: number;
}

export interface LearningPathDetail {
  id: string;
  title: string;
  description: string | null;
  image_url: string | null;
  pct: number;
  course_count: number;
  mission_count: number;
  total_duration_seconds: number;
  steps: LearningPathStep[];
  price_cents: number | null;
  currency: string;
  fully_owned: boolean;
}

// ── completion certificates (2026-08-13) ──────────────────────────────────

export interface MyCertificate {
  id: string;
  type: "lms_course_completion" | "lms_path_completion";
  /** The course or learning path it was earned for. */
  title: string;
  course_id: string | null;
  learning_path_id: string | null;
  issued_at: string | null;
  /** Signed at query time — don't cache this beyond the request. */
  url: string | null;
}

export async function fetchMyCertificates(): Promise<MyCertificate[]> {
  const { data } = await api.get<MyCertificate[]>("/lms/my-certificates");
  return data;
}

export async function fetchLearningPaths(): Promise<LearningPathCatalogItem[]> {
  const { data } = await api.get<LearningPathCatalogItem[]>("/lms/learning-paths");
  return data;
}

export async function fetchLearningPath(pathId: string): Promise<LearningPathDetail> {
  const { data } = await api.get<LearningPathDetail>(`/lms/learning-paths/${pathId}`);
  return data;
}

export async function startLearningPath(pathId: string): Promise<LearningPathDetail> {
  const { data } = await api.post<LearningPathDetail>(`/lms/learning-paths/${pathId}/start`);
  return data;
}

// ── upcoming public programs (reuses /public/catalog — no LMS-specific
// backend needed). Includes both `planned` (not yet open — "Notify me") and
// `registration_open` ("Register now") public cohorts as of 2026-08-07. ───

export interface UpcomingProgramSession {
  meeting_date: string;
  starts_at: string | null;
  title: string | null;
}

export interface UpcomingProgram {
  cohort_id: string;
  program_name: string;
  program_type: string;
  description: string | null;
  starts_on: string | null;
  ends_on: string | null;
  location: string | null;
  location_name: string | null;
  location_address: string | null;
  location_maps_url: string | null;
  price_display: string;
  capacity: number | null;
  spots_left: number | null;
  is_limited: boolean;
  registration_endpoint: string;
  status: "planned" | "registration_open";
  interest_endpoint: string;
  sessions: UpcomingProgramSession[];
  instructors: string[];
  curriculum_titles: string[];
}

export async function fetchUpcomingPrograms(): Promise<UpcomingProgram[]> {
  const { data } = await api.get<UpcomingProgram[]>("/public/catalog");
  return data;
}

// ── cities (2026-08-08) — no-auth read for pre-account forms (apply,
// signup) that can't call the authenticated /inventory/cities. ───────────
export interface PublicCity {
  id: string;
  name: string;
  country: string;
  is_active: boolean;
  created_at: string | null;
}

export async function fetchPublicCities(): Promise<PublicCity[]> {
  const { data } = await api.get<PublicCity[]>("/public/cities");
  return data;
}

export interface PublicLeadInput {
  student_name: string;
  email: string;
  phone: string;
}

// Full registration — mirrors PublicRegistrationRequest field-for-field
// (backend/app/schemas/sessions/public_registration.py). Only student_name/
// email/phone are required; everything else is optional, matching the
// backend schema exactly.
export interface ProgramRegistrationInput {
  student_name: string;
  email: string;
  phone: string;
  city?: string;
  date_of_birth?: string; // YYYY-MM-DD
  grade?: string;
  organization_name?: string;
  parent_name?: string;
  parent_phone?: string;
  parent_email?: string;
}

export async function submitProgramRegistration(endpoint: string, body: ProgramRegistrationInput): Promise<{ message: string }> {
  const { data } = await api.post<{ message: string }>(endpoint, body);
  return data;
}

export async function submitProgramInterest(endpoint: string, body: PublicLeadInput): Promise<{ message: string }> {
  const { data } = await api.post<{ message: string }>(endpoint, body);
  return data;
}

export async function fetchCatalog(q?: string): Promise<CourseCatalogItem[]> {
  const { data } = await api.get<CourseCatalogItem[]>("/lms/catalog", { params: q ? { q } : undefined });
  return data;
}

export async function fetchCourse(courseId: string): Promise<CourseDetail> {
  const { data } = await api.get<CourseDetail>(`/lms/courses/${courseId}`);
  return data;
}

export async function enrollInCourse(courseId: string): Promise<void> {
  await api.post("/lms/enroll", { course_id: courseId });
}

// Stage S (Stripe Checkout, August Build Brief Branch 4).

export async function startCourseCheckout(courseId: string): Promise<{ checkout_url: string }> {
  const { data } = await api.post<{ checkout_url: string }>(`/lms/courses/${courseId}/checkout`, {});
  return data;
}

// Learning path bundle pricing (2026-08-21).
export async function startPathCheckout(pathId: string): Promise<{ checkout_url: string }> {
  const { data } = await api.post<{ checkout_url: string }>(`/lms/learning-paths/${pathId}/checkout`, {});
  return data;
}

export interface CheckoutFulfillResult {
  status: "pending" | "paid" | "refunded" | "disputed" | "failed";
  course_id: string | null;
  learning_path_id: string | null;
}

export async function fulfillCheckoutSession(sessionId: string): Promise<CheckoutFulfillResult> {
  const { data } = await api.post<CheckoutFulfillResult>(`/lms/checkout/session/${sessionId}/fulfill`, {});
  return data;
}

export async function fetchModule(moduleId: string): Promise<ModuleDetail> {
  const { data } = await api.get<ModuleDetail>(`/lms/modules/${moduleId}`);
  return data;
}

export async function submitQuiz(itemId: string, answers: number[]): Promise<QuizReview> {
  const { data } = await api.post<QuizReview>(`/lms/items/${itemId}/quiz/submit`, { answers });
  return data;
}

export async function checkQuizAnswer(itemId: string, questionIndex: number, answer: number): Promise<QuizAnswerCheck> {
  const { data } = await api.post<QuizAnswerCheck>(`/lms/items/${itemId}/quiz/check`, { question_index: questionIndex, answer });
  return data;
}

export interface AttachmentUrl {
  url: string;
  filename: string | null;
}

export async function getAttachmentUrl(itemId: string): Promise<AttachmentUrl> {
  const { data } = await api.get<AttachmentUrl>(`/lms/items/${itemId}/attachment/url`);
  return data;
}

export type ProgressAction = "video-watched" | "text-viewed" | "quiz-attempt" | "flashcards-skipped" | "attachment-viewed";

export async function recordProgress(itemId: string, action: ProgressAction): Promise<ProgressResult> {
  const { data } = await api.post<ProgressResult>(`/lms/items/${itemId}/progress`, { action });
  return data;
}

export async function fetchVideoToken(itemId: string): Promise<{ token: string; expires_in_seconds: number }> {
  const { data } = await api.get<{ token: string; expires_in_seconds: number }>(
    `/lms/items/${itemId}/video/token`,
  );
  return data;
}

/** Absolute URL to the token-gated HLS playlist — hls.js loads this directly
 * (not through the `api` axios instance, which is for JSON calls). */
export function videoPlaylistUrl(itemId: string, token: string): string {
  const base = api.defaults.baseURL ?? "";
  return `${base}/lms/videos/${itemId}/playlist?token=${encodeURIComponent(token)}`;
}

// Leaderboard (P2-4, linked into the frontend in Live Games Phase 2C 8-2 —
// display_name is the student's nickname, never a real name).
export interface LeaderboardEntry {
  rank: number;
  user_id: string;
  display_name: string;
  points: number;
}

export const getLeaderboardApi = (scope: "cohort" | "global", cohortId?: string) =>
  api.get<LeaderboardEntry[]>("/lms/leaderboard", { params: { scope, cohort_id: cohortId } })
    .then((r) => r.data);

export async function fetchCheckpoints(videoItemId: string): Promise<VideoCheckpoint[]> {
  const { data } = await api.get<VideoCheckpoint[]>(`/lms/items/${videoItemId}/checkpoints`);
  return data;
}

export async function submitCheckpointAnswer(
  videoItemId: string,
  checkpointId: string,
  answer: number | number[] | string,
): Promise<CheckpointAnswerResult> {
  const { data } = await api.post<CheckpointAnswerResult>(
    `/lms/items/${videoItemId}/checkpoints/${checkpointId}/answer`, { answer },
  );
  return data;
}

// ── LMS Program checklist (2026-08-21 redesign) ─────────────────────────────
// Distinct from the older /lms/my-programs (courses-only cohort view, still
// live — LearnMyCourses/etc.) and from LearnProgram.tsx's /learn/programs/
// $cohortId (the *public catalog's* program detail/registration page, an
// unrelated concept that happens to share the word "program") — this is the
// new checklist entity, `lms_program_assignments`. Mounted at
// /learn/checklists to keep well clear of that existing route.

export type LmsProgramItemType = "course" | "mission_run" | "external_link" | "submission" | "article" | "manual";
export type LmsProgramItemStatus = "pending" | "done" | "awaiting_confirmation";

export interface LmsProgramChecklistItem {
  id: string;
  position: number;
  item_type: LmsProgramItemType;
  title: string;
  description: string | null;
  optional: boolean;
  requires_confirmation: boolean;
  status: LmsProgramItemStatus;
  course_id: string | null;
  mission_attempt_id: string | null;
  mission_id: string | null;
  mission_kind: string | null;
  external_url: string | null;
  submission_prompt: string | null;
  submitted_url: string | null;
}

export interface LmsProgramAssignmentSummary {
  assignment_id: string;
  lms_program_id: string;
  name: string;
  cohort_id: string | null;
  cohort_name: string | null;
  items_total: number;
  items_done: number;
  pct: number;
  next_item_title: string | null;
  certificate_required: boolean;
  certificate_earned: boolean;
}

export interface LmsProgramChecklist {
  assignment_id: string;
  lms_program_id: string;
  name: string;
  description: string | null;
  cohort_id: string | null;
  cohort_name: string | null;
  certificate_required: boolean;
  certificate_earned: boolean;
  items: LmsProgramChecklistItem[];
}

export async function fetchMyChecklists(): Promise<LmsProgramAssignmentSummary[]> {
  const { data } = await api.get<LmsProgramAssignmentSummary[]>("/lms/programs");
  return data;
}

export async function fetchChecklist(assignmentId: string): Promise<LmsProgramChecklist> {
  const { data } = await api.get<LmsProgramChecklist>(`/lms/programs/${assignmentId}`);
  return data;
}

export async function completeChecklistItem(assignmentId: string, itemId: string): Promise<LmsProgramChecklistItem> {
  const { data } = await api.post<LmsProgramChecklistItem>(`/lms/programs/${assignmentId}/items/${itemId}/complete`, {});
  return data;
}

export async function submitChecklistItem(assignmentId: string, itemId: string, url: string): Promise<LmsProgramChecklistItem> {
  const { data } = await api.post<LmsProgramChecklistItem>(`/lms/programs/${assignmentId}/items/${itemId}/submit`, { url });
  return data;
}
